CIPs/0005.md

```diff
@@ -0,0 +1,74 @@
+
+The asymmetric ignorance case can then be handled by having the validators directly communicate to each remote validator it's assigned enodeURL whenever the enodeURL mapping is updated.  Note that this will need to happen for all remote validators, even if the assigned URL is not updated for a subset of the validators, since all validators will need to know what is the updated version number.  Also, this direct communication will need to be handled when there is and is not an existing connection between validators.
+
+The mutual ignorance case can be handled by having each validator monitor when a validator enode mapping changes, and when it realizes that it's changed, it will gossip a query message to the validator for the updated assigned enodeURL.  That validator will then directly communicate it to the querying validator.
+
+### Direct communication of assigned enodeURL
+
+This component will include the following changes:
+
+1)  Announce message format will need to be changed so that only one authenticated and encrypted enodeURL (along with the version number) can be put in the message payload
+2)  On enodeURL mapping update, that validator will need to directly send an announce message to all validators. In the case of when the validator already has a connection with the remote validator, the originating validator should simply send an announce message to it.  In the case when that connection doesn't exist, then the announce message will need to be sent as part of the connection establishing handshake.
+
+Issue https://github.com/celo-org/celo-monorepo/issues/2667 tracks this component.
+
+### Fast discovery of changed enodeURL mapping
+
+This component will require all full nodes and validators to store a copy of a table that will contain the latest enodeURL mapping version number for each elected validator.  Whenver a validator updates it's enodeURL mapping, it will need to gossip a message containing the updated version number, and all nodes should update their table.
+
+Whenever a connection is established between two nodes, they will synchronize their tables so that both tables will have the latest version numbers.  This will handle the case of when a node just joins the P2P network and when a network partition has been repaired.  In those cases, the nodes throughout the network should quickly synchronize their tables.
+
+Issue https://github.com/celo-org/celo-monorepo/issues/2668 tracks this component.
+
+### EnodeURL query message
+
+When a validator realizes that it's assigned enodeURL for a remote validator has been updated, that validator should gossip a enodeURL query message destined to the remote validator.  Once the destined validator receives the message, it will send a direct announce message to the originating validator.
```

Comment:
How does the origin validator realize that?

Reply:
What's the benefit of this component?

Reply:
> How does the origin validator realize that?

Via the gossiped version number msg described in the above subsection.

> What's the benefit of this component?

This component is actually necessary.  Once a validator realizes that it's assigned enodeURL from a remote validator has potentially changed.  Then the origin validator will need to gossip this message.